File: deck.py
"""Deck management and shuffling for Flip 7."""

import random
from typing import Callable, List, Optional

from card import (
    Card,
    CardType,
    ActionType,
    ModifierType,
)


class Deck:
    """The game deck."""

    def __init__(self) -> None:
        self._cards: List[Card] = []
        self._discards: List[Card] = []
        self._rng = random.Random()
        self._debug_mode = False
        self._scanner = None  # Set via set_debug_mode; type: Optional[any]
        self.original_total = 0
        self._create_cards()
        self.shuffle()
        self.original_total = len(self._cards)

    def _create_cards(self) -> None:
        """Build all cards with correct distribution."""
        # Number cards: each number has as many cards as its value
        for value in range(13):  # 0..12
            count = value if value > 0 else 1
            for _ in range(count):
                self._cards.append(Card.new_number_card(value))

        # Score modifier cards (6 total)
        for mod in (
            ModifierType.PLUS_2,
            ModifierType.PLUS_4,
            ModifierType.PLUS_6,
            ModifierType.PLUS_8,
            ModifierType.PLUS_10,
            ModifierType.MULTIPLY_2,
        ):
            self._cards.append(Card.new_modifier_card(mod))

        # Action cards (3 of each type = 9 total)
        for _ in range(3):
            for action in (ActionType.FREEZE, ActionType.FLIP_THREE, ActionType.SECOND_CHANCE):
                self._cards.append(Card.new_action_card(action))

    def shuffle(self) -> None:
        self._rng.shuffle(self._cards)

    def draw_card(self) -> Card:
        if not self._cards:
            raise RuntimeError("All cards disappeared!")
        if self._debug_mode:
            return self._draw_card_debug()

        card = self._cards.pop()
        if not self._cards:
            self._reshuffle()
        return card

    def discard_card(self, card: Optional[Card]) -> None:
        if card is not None:
            self._discards.append(card)

    def _reshuffle(self) -> None:
        self._cards.extend(self._discards)
        self._discards = []
        self.shuffle()

    def cards_left(self) -> int:
        return len(self._cards)

    def total_cards(self) -> int:
        return len(self._cards) + len(self._discards)

    def set_debug_mode(self, debug: bool, scanner: Optional[Callable[[], str]] = None) -> None:
        self._debug_mode = debug
        self._scanner = scanner

    def _draw_card_debug(self) -> Card:
        if not self._cards:
            raise RuntimeError("No cards in deck")
        number_cards: List[Card] = []
        action_cards: List[Card] = []
        modifier_cards: List[Card] = []
        for c in self._cards:
            if c.type == CardType.NUMBER:
                number_cards.append(c)
            elif c.type == CardType.ACTION:
                action_cards.append(c)
            else:
                modifier_cards.append(c)

        card_options: List[Card] = []
        option_index = 1

        print("\n🐛 DEBUG: Choose a card to draw:")
        print(f"Available cards ({len(self._cards)} total):")

        if number_cards:
            print("\nNumber Cards:")
            card_counts: dict[int, int] = {}
            for c in number_cards:
                card_counts[c.value] = card_counts.get(c.value, 0) + 1
            for value in range(13):
                count = card_counts.get(value, 0)
                if count > 0:
                    print(f"  {option_index}) [{value}] ({count} available)")
                    card_options.append(Card.new_number_card(value))
                    option_index += 1

        if action_cards:
            print("\nAction Cards:")
            action_names = ["❄️ FREEZE", "🎲 FLIP 3", "🆘 2ND CHANCE"]
            for i, at in enumerate([ActionType.FREEZE, ActionType.FLIP_THREE, ActionType.SECOND_CHANCE]):
                count = sum(1 for c in action_cards if c.action == at)
                if count > 0:
                    print(f"  {option_index}) {action_names[i]} ({count} available)")
                    card_options.append(Card.new_action_card(at))
                    option_index += 1

        if modifier_cards:
            print("\nModifier Cards:")
            mod_names = ["+2", "+4", "+6", "+8", "+10", "×2"]
            for i, mt in enumerate(ModifierType):
                count = sum(1 for c in modifier_cards if c.modifier == mt)
                if count > 0:
                    print(f"  {option_index}) [{mod_names[i]}] ({count} available)")
                    card_options.append(Card.new_modifier_card(mt))
                    option_index += 1

        n = len(card_options)
        print(f"\nEnter choice (1-{n}): ", end="", flush=True)

        while True:
            if self._scanner is None:
                return self._draw_random_card()
            try:
                line = self._scanner()
            except Exception:
                return self._draw_random_card()
            line = (line or "").strip()
            try:
                choice = int(line)
            except ValueError:
                print(f"Please enter a number between 1 and {n}: ", end="", flush=True)
                continue
            if choice < 1 or choice > n:
                print(f"Please enter a number between 1 and {n}: ", end="", flush=True)
                continue

            selected = card_options[choice - 1]
            for i, c in enumerate(self._cards):
                if self._cards_equal(c, selected):
                    return self._cards.pop(i)
            print("Card not found, drawing random card instead...")
            return self._draw_random_card()

    def _draw_random_card(self) -> Card:
        if not self._cards:
            raise RuntimeError("No cards in deck")
        return self._cards.pop()

    def _cards_equal(self, c1: Card, c2: Card) -> bool:
        if c1.type != c2.type:
            return False
        if c1.type == CardType.NUMBER:
            return c1.value == c2.value
        if c1.type == CardType.ACTION:
            return c1.action == c2.action
        if c1.type == CardType.MODIFIER:
            return c1.modifier == c2.modifier
        return False

    # Expose for game's card count checks
    @property
    def cards(self) -> List[Card]:
        return self._cards

    @property
    def discards(self) -> List[Card]:
        return self._discards

File: go/deck.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"strconv"
	"strings"
	"time"
)

// Deck represents the game deck
type Deck struct {
	cards         []*Card
	discards      []*Card
	rng           *rand.Rand
	debugMode     bool
	scanner       *bufio.Scanner
	OriginalTotal int
}

// NewDeck creates a new deck with the correct card distribution for Flip 7
func NewDeck() *Deck {
	deck := &Deck{
		cards:    make([]*Card, 0),
		discards: make([]*Card, 0),
		rng:      rand.New(rand.NewSource(time.Now().UnixNano())),
	}

	deck.createCards()
	deck.Shuffle()
	deck.OriginalTotal = len(deck.cards)

	return deck
}

// createCards creates all cards with the correct distributions
func (d *Deck) createCards() {
	// Number cards: each number has as many cards as its value
	// 12 has 12 copies, 11 has 11 copies, etc., down to 0 which has 1 copy
	for value := 0; value <= 12; value++ {
		count := value
		if value == 0 {
			count = 1
		}
		for i := 0; i < count; i++ {
			d.cards = append(d.cards, NewNumberCard(value))
		}
	}

	// Score Modifier Cards (6 total)
	d.cards = append(d.cards, NewModifierCard(Plus2))
	d.cards = append(d.cards, NewModifierCard(Plus4))
	d.cards = append(d.cards, NewModifierCard(Plus6))
	d.cards = append(d.cards, NewModifierCard(Plus8))
	d.cards = append(d.cards, NewModifierCard(Plus10))
	d.cards = append(d.cards, NewModifierCard(Multiply2))

	// Action Cards (3 of each type = 9 total)
	for i := 0; i < 3; i++ {
		d.cards = append(d.cards, NewActionCard(Freeze))
		d.cards = append(d.cards, NewActionCard(FlipThree))
		d.cards = append(d.cards, NewActionCard(SecondChance))
	}
}

// Shuffle shuffles the deck
func (d *Deck) Shuffle() {
	d.rng.Shuffle(len(d.cards), func(i, j int) {
		d.cards[i], d.cards[j] = d.cards[j], d.cards[i]
	})
}

// DrawCard draws the top card from the deck
func (d *Deck) DrawCard() *Card {
	if len(d.cards) == 0 {
		panic("All cards disappeared!	")
	}
	if d.debugMode {
		return d.drawCardDebug()
	}

	card := d.cards[len(d.cards)-1]
	d.cards = d.cards[:len(d.cards)-1]

	if len(d.cards) == 0 {
		d.Reshuffle()
	}

	return card
}

// DiscardCard adds a card to the discard pile
func (d *Deck) DiscardCard(card *Card) {
	if card != nil {
		d.discards = append(d.discards, card)
	}
}

// Reshuffle reshuffles the discard pile back into the deck
func (d *Deck) Reshuffle() {
	d.cards = append(d.cards, d.discards...)
	d.discards = make([]*Card, 0)
	d.Shuffle()
}

// CardsLeft returns the number of cards remaining in the deck
func (d *Deck) CardsLeft() int {
	return len(d.cards)
}

// TotalCards returns the total number of cards (deck + discards)
func (d *Deck) TotalCards() int {
	return len(d.cards) + len(d.discards)
}

// SetDebugMode enables or disables debug mode for manual card selection
func (d *Deck) SetDebugMode(debug bool, scanner *bufio.Scanner) {
	d.debugMode = debug
	d.scanner = scanner
}

// drawCardDebug allows manual selection of cards in debug mode
func (d *Deck) drawCardDebug() *Card {
	if len(d.cards) == 0 {
		return nil
	}

	fmt.Println("\n🐛 DEBUG: Choose a card to draw:")
	fmt.Printf("Available cards (%d total):\n", len(d.cards))

	// Group cards by type for easier selection
	numberCards := make([]*Card, 0)
	actionCards := make([]*Card, 0)
	modifierCards := make([]*Card, 0)

	for _, card := range d.cards {
		switch card.Type {
		case NumberCard:
			numberCards = append(numberCards, card)
		case ActionCard:
			actionCards = append(actionCards, card)
		case ModifierCard:
			modifierCards = append(modifierCards, card)
		}
	}

	// Display cards by category
	cardOptions := make([]*Card, 0)
	optionIndex := 1

	if len(numberCards) > 0 {
		fmt.Println("\nNumber Cards:")
		cardCounts := make(map[int]int)
		for _, card := range numberCards {
			cardCounts[card.Value]++
		}
		for value := 0; value <= 12; value++ {
			if count := cardCounts[value]; count > 0 {
				fmt.Printf("  %d) [%d] (%d available)\n", optionIndex, value, count)
				cardOptions = append(cardOptions, NewNumberCard(value))
				optionIndex++
			}
		}
	}

	if len(actionCards) > 0 {
		fmt.Println("\nAction Cards:")
		actionCounts := make(map[ActionType]int)
		for _, card := range actionCards {
			actionCounts[card.Action]++
		}
		actionNames := []string{"❄️ FREEZE", "🎲 FLIP 3", "🆘 2ND CHANCE"}
		for i, count := range []int{actionCounts[Freeze], actionCounts[FlipThree], actionCounts[SecondChance]} {
			if count > 0 {
				fmt.Printf("  %d) %s (%d available)\n", optionIndex, actionNames[i], count)
				cardOptions = append(cardOptions, NewActionCard(ActionType(i)))
				optionIndex++
			}
		}
	}

	if len(modifierCards) > 0 {
		fmt.Println("\nModifier Cards:")
		modifierCounts := make(map[ModifierType]int)
		for _, card := range modifierCards {
			modifierCounts[card.Modifier]++
		}
		modifierNames := []string{"+2", "+4", "+6", "+8", "+10", "×2"}
		for i, count := range []int{modifierCounts[Plus2], modifierCounts[Plus4], modifierCounts[Plus6], modifierCounts[Plus8], modifierCounts[Plus10], modifierCounts[Multiply2]} {
			if count > 0 {
				fmt.Printf("  %d) [%s] (%d available)\n", optionIndex, modifierNames[i], count)
				cardOptions = append(cardOptions, NewModifierCard(ModifierType(i)))
				optionIndex++
			}
		}
	}

	fmt.Printf("\nEnter choice (1-%d): ", len(cardOptions))

	for {
		if !d.scanner.Scan() {
			// Fall back to random card if input fails
			return d.drawRandomCard()
		}

		input := strings.TrimSpace(d.scanner.Text())
		choice, err := strconv.Atoi(input)
		if err != nil || choice < 1 || choice > len(cardOptions) {
			fmt.Printf("Please enter a number between 1 and %d: ", len(cardOptions))
			continue
		}

		selectedCard := cardOptions[choice-1]

		// Find and remove the actual card from the deck
		for i, card := range d.cards {
			if d.cardsEqual(card, selectedCard) {
				// Remove this card from the deck
				d.cards = append(d.cards[:i], d.cards[i+1:]...)
				return card
			}
		}

		// Fallback if card not found (shouldn't happen)
		fmt.Println("Card not found, drawing random card instead...")
		return d.drawRandomCard()
	}
}

// drawRandomCard draws a random card (fallback method)
func (d *Deck) drawRandomCard() *Card {
	if len(d.cards) == 0 {
		return nil
	}

	card := d.cards[len(d.cards)-1]
	d.cards = d.cards[:len(d.cards)-1]
	return card
}

// cardsEqual checks if two cards are equivalent
func (d *Deck) cardsEqual(card1, card2 *Card) bool {
	if card1.Type != card2.Type {
		return false
	}

	switch card1.Type {
	case NumberCard:
		return card1.Value == card2.Value
	case ActionCard:
		return card1.Action == card2.Action
	case ModifierCard:
		return card1.Modifier == card2.Modifier
	}

	return false
}
